### PR TITLE
Lang mapper

### PR DESCRIPTION
**Pull request by phlax**
Adds a language mapper with presets for Projects

At the moment its not hooked up (other than in pootle_fs), but im thinking that Store.update/sync should also respect any configured lang mappings.

initial docs here https://github.com/translate/pootle/wiki/DocsDevLangMapper

### CODE REVIEW

File: pootle/apps/pootle_project/lang_mapper.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    def _parse_mappings(self):
+        self._mapping = OrderedDict()
+
+        for k, v in self.project_mapping.items():
+            if k.startswith("$"):
+                preset = self.site_presets.get(v, None)
+                if not preset:
+                    logger.warning(
+                        "Unrecognised lang mapping preset: %s" % preset)
+                else:
```
Comment (unho) [inline pootle/apps/pootle_project/lang_mapper.py]: Nit: We can avoid the `else` by using `continue`.

File: pootle/apps/pootle_project/lang_mapper.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            return Language.objects.get(
+                code=self.get_pootle_code(lang_code))
+        except Language.DoesNotExist:
+            return None
+
+    def get_pootle_code(self, lang_code):
+        return self.lang_mappings.get(lang_code, lang_code)
+
+    def _add_lang_to_mapping(self, k, v):
+        # as its a 1-2-1 mapping remove any previous items with
```
Comment (unho) [inline pootle/apps/pootle_project/lang_mapper.py]: "1 to 1"?

File: pootle/apps/pootle_project/lang_mapper.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    @lru_cache()
+    def get_lang(self, lang_code):
+        try:
+            return Language.objects.get(
+                code=self.get_pootle_code(lang_code))
+        except Language.DoesNotExist:
+            return None
+
+    def get_pootle_code(self, lang_code):
```
Comment (unho) [inline pootle/apps/pootle_project/lang_mapper.py]: Nit: I would put `get_pootle_code` and `get_upstream_code` one after the other.

File: pootle/apps/pootle_project/lang_mapper.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            for _k, _v in self._mapping.items():
+                if v == _v:
+                    del self._mapping[_k]
+                    break
+        self._mapping[k] = v
+
+    def _parse_mappings(self):
+        self._mapping = OrderedDict()
+
+        for k, v in self.project_mapping.items():
```
Comment (unho) [inline pootle/apps/pootle_project/lang_mapper.py]: `k` and `v` do not convey any information. But I see that above `fs_code` and `pootle_code` are used, and those are easier to understand.

File: pootle/apps/pootle_project/lang_mapper.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+from pootle_language.models import Language
+
+
+logger = logging.getLogger(__name__)
+
+
+class LanguageMapper(object):
+    """For a given code find the relevant pootle lang taking account of any
+    lang mapping configuration
+
```
Comment (unho) [inline pootle/apps/pootle_project/lang_mapper.py]: Unneeded blank line.

**Comment by unho** [comment]
I guess that the purpose of this code is to match Pootle language code (for example `ne`) to the code used in the filesystem (for example `ne_NP`), isn't it? I am asking because I don't feel this is explained at all. It would be nice to have at least some docstrings explaining it.

**Comment by phlax** [comment]
@unho you are right on both accounts.

As this is configured with `pootle config` and currently is not hooked into anything (other than pootle_fs) i would like to land as is, with some docs and a PR to hook in properly to follow shortly after.

adding some docstrings now...

File: pootle/apps/pootle_project/lang_mapper.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                preset = self.site_presets.get(pootle_code, None)
+                if not preset:
+                    logger.warning(
+                        "Unrecognised lang mapping preset: %s" % preset)
+                    continue
+                for preset_upstream_code, preset_pootle_code in preset:
+                    self._add_lang_to_mapping(
+                        preset_upstream_code,
+                        preset_pootle_code)
+            else:
```
Comment (unho) [inline pootle/apps/pootle_project/lang_mapper.py]: This `else` can also be avoided by using `continue`.

**Comment by unho** [comment]
I checked docstrings and I still don't see any real example. I think a real example might help understanding the purpose.

File: pootle/apps/pootle_project/lang_mapper.py (hunk 6 above)
Comment (phlax) [inline pootle/apps/pootle_project/lang_mapper.py]: im not so sure that is more readable tbh

File: pootle/apps/pootle_project/lang_mapper.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+from pootle.core.delegate import config
+from pootle_language.models import Language
+
+
+logger = logging.getLogger(__name__)
+
+
+class LanguageMapper(object):
+    """For a given code find the relevant pootle lang taking account of any
```
Comment (unho) [inline pootle/apps/pootle_project/lang_mapper.py]: Pootle language

File: pootle/apps/pootle_project/lang_mapper.py (hunk 6 above)
Comment (unho) [inline pootle/apps/pootle_project/lang_mapper.py]: k

**Comment by unho** [comment]
lgtm

File: tests/misc/lang_mapper.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+     ["bar",
+      [["testbar1", "en"],
+       ["testbar2", "language1"],
+       ["testbar3", "language0"],
+       ["testbar4", "bar"]]],
+     ["baz",
+      [["testbaz1", "en"],
+       ["testbaz2", "language1"],
+       ["testbaz3", "language0"],
+       ["testbaz4", "baz"]]]])
```
Comment (dwaynebailey) [inline tests/misc/lang_mapper.py]: foo, bar, baz makes it frustratingly hard to understand what exactly is being tested.  Real codes for real languages would have helped way way more.

Comment (phlax) [inline tests/misc/lang_mapper.py]: ill rename test[foo|bar|baz] to "upstream_code_n"

Comment (phlax) [inline tests/misc/lang_mapper.py]: and for the preset names `preset_n`

**Comment by ta2-1** [comment]
@phlax great work. thanks. Looks very good after you added `pootle.core.use_lang_mapping_presets`. GTM.

**Pull request merged**